futures-util/src/stream/mapped_futures/mod.rs

```diff
@@ -0,0 +1,464 @@
+// inserted into the ready to run queue, so that its future can be polled later.
+//
+// Each task is wrapped in an `Arc` and thereby atomically reference counted.
+// Also, each task contains an `AtomicBool` which acts as a flag that indicates
+// whether the task is currently inserted in the atomic queue. When a wake-up
+// notification is received, the task will only be inserted into the ready to
+// run queue if it isn't inserted already.
+
+impl<K: Hash + Eq, Fut> Default for MappedFutures<K, Fut> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<K: Hash + Eq, Fut> MappedFutures<K, Fut> {
+    /// Constructs a new, empty [`MappedFutures`].
+    ///
+    /// The returned [`MappedFutures`] does not contain any futures.
+    /// In this state, [`MappedFutures::poll_next`](Stream::poll_next) will
+    /// return [`Poll::Ready(None)`](Poll::Ready).
+    pub fn new() -> Self {
+        Self { inner: FuturesUnorderedInternal::new() }
+    }
+
+    /// Returns the number of futures contained in the set.
+    ///
+    /// This represents the total number of in-flight futures.
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    /// Returns `true` if the set contains no futures.
+    pub fn is_empty(&self) -> bool {
+        // Relaxed ordering can be used here since we don't need to read from
+        // the head pointer, only check whether it is null.
+        self.inner.is_empty()
+    }
+
+    /// This is a code stank
+    fn set(&mut self) -> &mut HashSet<HashTask<K, Fut>> {
+        &mut self.inner.inner.inner
+    }
+
+    /// Insert a future into the set.
+    ///
+    /// This method adds the given future to the set. This method will not
+    /// call [`poll`](core::future::Future::poll) on the submitted future. The caller must
+    /// ensure that [`MappedFutures::poll_next`](Stream::poll_next) is called
+    /// in order to receive wake-up notifications for the given future.
+    ///
+    /// This method will remove and drop a future that is already mapped to the provided key.
+    /// Returns true if another future was not removed to make room for the provided future.
+    pub fn insert(&mut self, key: K, future: Fut) -> bool {
+        let replacing = self.cancel(&key);
+        let task = self.inner.push(key, future);
+        // self.inner.self.hash_set.insert(task.into());
+        self.set().insert(task.into());
+        !replacing
+    }
+
+    /// Insert a future into the set and return the displaced future, if there was one.
+    ///
+    /// This method adds the given future to the set. This method will not
+    /// call [`poll`](core::future::Future::poll) on the submitted future. The caller must
+    /// ensure that [`MappedFutures::poll_next`](Stream::poll_next) is called
+    /// in order to receive wake-up notifications for the given future.
+    /// Returns true if another future was ma
+    pub fn replace(&mut self, key: K, future: Fut) -> Option<Fut>
```

Comment:
Having both `replace` and `insert` (and remove + cancel below) seems redundant with the only difference seems to be if the value is dropped before the function returns or not. std's maps only have `insert`, which behaves as `replace` here, have you considered just having `insert` with `replace`'s signature?

Reply:
The reasoning for having the separate methods is covered in the PR, nonetheless I'm open to ideas for how these functions could be renamed to make their intention more clear. Ideally each should be a single word, and the result, and difference between the methods should be clear at a glance, without reading the full signature. For remove & cancel it is clear, since the mental model of removing something includes that you have a thing after doing so, and not so with cancel. But is the difference as clear with replace & insert?